lib/services/app.go

```
@@ -158,3 +169,217 @@ func UnmarshalAppServer(data []byte, opts ...MarshalOption) (types.AppServer, er
+	constructURI := func(p string) string { return fmt.Sprintf("%s://%s:%d", p, serviceFQDN, port.Port) }
+
+	proto := strings.ToLower(protocolAnnotation)
+	if proto == protoHTTP || proto == protoHTTPS {
+		return constructURI(proto)
+	}
+
+	if port.AppProtocol != nil {
+		switch strings.ToLower(*port.AppProtocol) {
+		case protoHTTP:
+			return constructURI(protoHTTP)
+		case protoHTTPS:
+			return constructURI(protoHTTPS)
+		}
+	}
+
+	if port.Port == 443 || strings.ToLower(port.Name) == protoHTTPS {
+		return constructURI(protoHTTPS)
+	}
+
+	if pc != nil {
+		result := pc.CheckProtocol(fmt.Sprintf("%s:%d", serviceFQDN, port.Port))
+		if result != protoTCP {
+			return constructURI(result)
+		}
+	}
+
+	if port.Port == 80 || port.Port == 8080 || strings.ToLower(port.Name) == protoHTTP {
+		return constructURI(protoHTTP)
+	}
+
+	return constructURI(protoTCP)
+}
+
+func getAppRewriteConfig(annotations map[string]string) (*types.Rewrite, error) {
+	rewrite := annotations[types.DiscoveryAppRewriteLabel]
+	if rewrite == "" {
+		return nil, nil
+	}
+
+	rw := types.Rewrite{}
+	reader := strings.NewReader(rewrite)
+	decoder := kyaml.NewYAMLOrJSONDecoder(reader, 32*1024)
+	err := decoder.Decode(&rw)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed decoding rewrite config")
+	}
+
+	return &rw, nil
+}
+
+func getAppName(serviceName, namespace, clusterName, portName string) string {
+	clusterName = strings.ReplaceAll(clusterName, ".", "-")
+	if portName != "" {
+		return fmt.Sprintf("%s-%s-%s-%s", serviceName, portName, namespace, clusterName)
+	}
+	return fmt.Sprintf("%s-%s-%s", serviceName, namespace, clusterName)
+}
+
+func getAppLabels(serviceLabels map[string]string, clusterName string) map[string]string {
+	result := make(map[string]string, len(serviceLabels)+1)
+
+	for k, v := range serviceLabels {
+		if !types.IsValidLabelKey(k) {
+			logrus.Debugf("Skipping label %q as invalid while creating app labels from service", k)
```

use the configured logger since it has the preferences of the user

---

I think this `IsValidLabelKey()` is not required, since what teleport allows in label keys is actually a superset of what Kubernetes allows. So I'm inclined to remove it.

---

I return error now instead of just skipping label.